extensions/amp-brightcove/1.0/component.js

```diff
@@ -0,0 +1,225 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '#preact';
+import {addParamsToUrl} from '../../../src/url';
+import {useCallback, useMemo, useRef, useState} from '#preact';
+import {forwardRef} from '#preact/compat';
+import {VideoIframe} from '../../amp-video/1.0/video-iframe';
+import {mutedOrUnmutedEvent} from '../../../src/iframe-video';
+import {VideoEvents} from '../../../src/video-interface';
+import {dispatchCustomEvent} from '#core/dom';
+
+/** @const {string} */
+const DEFAULT = 'default';
+
+/**
+ * @enum {string}
+ * @private
+ */
+const methods = {
+  'mute': 'muted',
+  'unmute': 'muted',
+  'showControls': 'controls',
+  'hideControls': 'controls',
+};
+
+/**
+ * @enum {string}
+ * @private
+ */
+const events = {
+  'ready': VideoEvents.LOAD,
+  'playing': VideoEvents.PLAYING,
+  'pause': VideoEvents.PAUSE,
+  'ended': VideoEvents.ENDED,
+  'ads-ad-started': VideoEvents.AD_START,
+  'ads-ad-ended': VideoEvents.AD_END,
+  'loadedmetadata': VideoEvents.LOADEDMETADATA,
+};
+
+/**
+ * @param {!BrightcoveDef.Props} props
+ * @param {{current: ?VideoWrapperDef.Api.CarouselApi}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function BrightcoveWithRef(
+  {
+    account,
+    autoplay,
+    embed = DEFAULT,
+    muted: mutedProp,
+    onReady,
```

Comment:
Why are `onReady` and `onPlayingState` props? What do you expect to pass in?

You should *always* `dispatchEvent` for `canplay` (on ready) and `play`/`pause`, these are handled in the Preact tree. See `amp-vimeo` for an example.

Reply:
Removed `onReady` from AMP layer so potentially may not need it in Preact. What's the difference between calling `dispatchEvent` for `canplay` in the `onMessage` when `data.event === 'ready'` vs. calling it in the `onIframeLoad` given to `VideoIframe`?

`onPlayingState` is called when `isPlaying` changes. It looks like 0.1 uses `PauseHelper` so I figured this should be updated in the AMP layer. Correct me if I'm wrong:
https://github.com/ampproject/amphtml/blob/301af16b40760c81c74295fc01c09de76d3a72a8/extensions/amp-brightcove/0.1/amp-brightcove.js#L240-L248

Reply:
Discussed offline - `onIframeLoads` identifies when the document loads, `onMessage` more precisely identifies when the document will start honoring playback requests. Therefore `onMessage` is preferred`.